File: day5/go/part_2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Rule [2]int
type Update []int

func parseInput(filePath string) ([]Rule, []Update) {
	file, err := os.Open(filePath)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var rules []Rule
	var updates []Update
	section := 0

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			section++
			continue
		}

		if section == 0 {
			parts := strings.Split(line, "|")
			a, _ := strconv.Atoi(parts[0])
			b, _ := strconv.Atoi(parts[1])
			rules = append(rules, Rule{a, b})
		} else {
			var update Update
			for _, numStr := range strings.Split(line, ",") {
				num, _ := strconv.Atoi(numStr)
				update = append(update, num)
			}
			updates = append(updates, update)
		}
	}

	return rules, updates
}

func sortUpdate(update Update, rules []Rule) Update {
	dependencies := make(map[int]map[int]bool)
	for _, page := range update {
		dependencies[page] = make(map[int]bool)
	}

	for _, rule := range rules {
		if _, ok1 := dependencies[rule[0]]; ok1 {
			if _, ok2 := dependencies[rule[1]]; ok2 {
				dependencies[rule[1]][rule[0]] = true
			}
		}
	}

	visited := make(map[int]bool)
	var sortedUpdate Update

	var dfs func(int)
	dfs = func(page int) {
		if visited[page] {
			return
		}
		visited[page] = true
		for dep := range dependencies[page] {
			dfs(dep)
		}
		sortedUpdate = append(sortedUpdate, page)
	}

	for _, page := range update {
		dfs(page)
	}

	return sortedUpdate
}

func main() {
	rules, updates := parseInput("input.txt")
	total := 0

	for _, update := range updates {
		sortedUpdate := sortUpdate(update, rules)
		total += sortedUpdate[len(sortedUpdate)/2]
	}

	fmt.Println(total)
}

File: day5/rust/part_2.rs
use std::collections::{HashMap, HashSet};
use std::fs::read_to_string;

type Rule = (i32, i32);
type Update = Vec<i32>;

fn parse_input(file_path: &str) -> (Vec<Rule>, Vec<Update>) {
    let content = read_to_string(file_path).unwrap();
    let mut rules = Vec::new();
    let mut updates = Vec::new();
    let mut section = 0;

    for line in content.lines() {
        let line = line.trim();
        if line.is_empty() {
            section += 1;
            continue;
        }

        if section == 0 {
            let parts: Vec<i32> = line.split('|')
                .map(|s| s.parse().unwrap())
                .collect();
            rules.push((parts[0], parts[1]));
        } else {
            let update: Vec<i32> = line.split(',')
                .map(|s| s.parse().unwrap())
                .collect();
            updates.push(update);
        }
    }

    (rules, updates)
}

fn sort_update(update: &[i32], rules: &[Rule]) -> Vec<i32> {
    let mut dependencies: HashMap<i32, HashSet<i32>> = update.iter()
        .map(|&page| (page, HashSet::new()))
        .collect();

    for &(a, b) in rules {
        if dependencies.contains_key(&a) && dependencies.contains_key(&b) {
            dependencies.get_mut(&b).unwrap().insert(a);
        }
    }

    let mut sorted_update = Vec::new();
    let mut visited = HashSet::new();

    fn dfs(
        page: i32,
        visited: &mut HashSet<i32>,
        dependencies: &HashMap<i32, HashSet<i32>>,
        sorted_update: &mut Vec<i32>,
    ) {
        if visited.contains(&page) {
            return;
        }
        visited.insert(page);
        for &dep in dependencies.get(&page).unwrap() {
            dfs(dep, visited, dependencies, sorted_update);
        }
        sorted_update.push(page);
    }

    for &page in update {
        dfs(page, &mut visited, &dependencies, &mut sorted_update);
    }

    sorted_update
}

fn main() {
    let (rules, updates) = parse_input("input.txt");
    let total: i32 = updates.iter()
        .map(|update| sort_update(update, &rules))
        .map(|sorted| sorted[sorted.len() / 2])
        .sum();

    println!("{}", total);
}
